src/librustc_front/hir.rs

```diff
@@ -50,7 +50,59 @@ use print::pprust;
 use util;
 
 use std::fmt;
-use serialize::{Encodable, Encoder, Decoder};
+use std::hash::{Hash, Hasher};
+use serialize::{Encodable, Decodable, Encoder, Decoder};
+
+/// Identifier in HIR
+#[derive(Clone, Copy, Eq)]
+pub struct Ident {
+    /// Hygienic name (renamed), should be used by default
+    pub name: Name,
+    /// Unhygienic name (original, not renamed), needed in few places in name resolution
+    pub unhygienic_name: Name,
+}
+
+impl Ident {
+    pub fn from_name(name: Name) -> Ident {
```

Comment:
I'm not sure how to call it better. For example, `for` loop desugaring calls `from_name` with gensym `iter` exactly for `unhygienic_name` to be hygienic.